lib/proxy/peer/internal/tls.go

```diff
@@ -0,0 +1,85 @@
+// Teleport
+// Copyright (C) 2024 Gravitational, Inc.
+//
+// This program is free software: you can redistribute it and/or modify
+// it under the terms of the GNU Affero General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+//
+// This program is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU Affero General Public License for more details.
+//
+// You should have received a copy of the GNU Affero General Public License
+// along with this program.  If not, see <http://www.gnu.org/licenses/>.
+
+package internal
+
+import (
+	"crypto/x509"
+	"slices"
+
+	"github.com/gravitational/trace"
+
+	"github.com/gravitational/teleport/api/types"
+	"github.com/gravitational/teleport/lib/tlsca"
+)
+
+type verifyPeerCertificateFunc = func(rawCerts [][]byte, verifiedChains [][]*x509.Certificate) error
+
+var _ verifyPeerCertificateFunc = VerifyPeerCertificateIsProxy
+
+func VerifyPeerCertificateIsProxy(rawCerts [][]byte, verifiedChains [][]*x509.Certificate) error {
+	if len(verifiedChains) < 1 {
+		return trace.AccessDenied("missing client certificate (this is a bug)")
+	}
+
+	clientCert := verifiedChains[0][0]
+	clientIdentity, err := tlsca.FromSubject(clientCert.Subject, clientCert.NotAfter)
+	if err != nil {
+		return trace.Wrap(err)
+	}
+
+	if !slices.Contains(clientIdentity.Groups, string(types.RoleProxy)) {
+		return trace.AccessDenied("expected Proxy client credentials")
+	}
+	return nil
+}
+
+func VerifyPeerCertificateIsSpecificProxy(peerID string) verifyPeerCertificateFunc {
+	return func(rawCerts [][]byte, verifiedChains [][]*x509.Certificate) error {
+		if len(verifiedChains) < 1 {
+			return trace.AccessDenied("missing server certificate (this is a bug)")
+		}
+
+		clientCert := verifiedChains[0][0]
+		clientIdentity, err := tlsca.FromSubject(clientCert.Subject, clientCert.NotAfter)
+		if err != nil {
+			return trace.Wrap(err)
+		}
+
+		if !slices.Contains(clientIdentity.Groups, string(types.RoleProxy)) {
+			return trace.AccessDenied("expected Proxy server credentials")
+		}
+
+		if clientIdentity.Username != peerID {
+			return trace.Wrap(WrongProxyError{})
+		}
+		return nil
+	}
+}
+
+const DuplicatePeerMsg = "Detected multiple Proxy Peers with the same public address when connecting to a Proxy which can lead to inconsistent state and problems establishing sessions. For best results ensure that `peer_public_addr` is unique per proxy and not a load balancer."
+
+type WrongProxyError struct{}
```

Comment:
It's not plain string matching, it takes the AccessDeniedError type into account as well. It also matches your impl, as an errors.Is with that exact same instance still matches WrongProxyError because of your Unwrap.

Reply:
It will cease to match if we change the message somewhere but not somewhere else, and it will also match against a completely unrelated `trace.AccessDenied("connected to unexpected proxy")`.

Reply:
I did mean for you to keep it as a global var, and for other places to keep using it. I'm just not seeing what benefits the type brings.

This already matches an unrelated `trace.AccessDenied("connected to unexpected proxy")` error. It's worth asking if that's really a problem, or if they are indeed distinct if they match so exactly.

Reply:
Satisfying `trace.IsAccessDenied` is a property that I didn't want to break from the old code; the match is only in one direction, not the other:
```go
require.ErrorIs(t, WrongProxyError{}, trace.AccessDenied(WrongProxyError{}.Error()))
require.NotErrorIs(t, trace.AccessDenied(WrongProxyError{}.Error()), WrongProxyError{})
```

i.e. a `WrongProxyError` is a `*trace.AccessDeniedError`, but not viceversa.

Reply:
I still think this could be quite a ways simpler, but I'll leave it to you.